README.md
# Granica

Authentication and authorization service.

The final intention of this project is to build a complete authentication and authorization system that can be easily deployed as a microservice.

But at the same time it is also a way to find a modular and systematized foundation to build microservices and micromonoliths; therefore, as a pattern emerges some features, if not alli, will serve to update a series of libraries and a [command-line interface](https://gitlab.com/mikrowezel/backend/cli) that will allow to automate as much as possible the generation of Go-based services while trying not to detach too much from the particular idiosyncrasy of this platform.

<img src="docs/img/users_index.png" width="480">

[Screenshots](docs/screenshots.md)

## Dev branch

- [new/wip at GitLab](https://gitlab.com/mikrowezel/backend/granica/tree/new/wip)
- [new/wip at GitHub](https://github.com/adrianpk/granica/tree/new/wip)

## Changelog

- [20191221 - User signup confirmation processing](/docs/changelog.md#20191221)
- [20191219 - Mailer](/docs/changelog.md#20191219)
- [20191218 - User signup](/docs/changelog.md#20191218)
- [20191217 - User signin](/docs/changelog.md#20191217)
- [20191216 - Model validation and flash messages](/docs/changelog.md#20191216)
- [20191215 - Form validations](/docs/changelog.md#20191215)
- [20191213 - RESTful actions completed](/docs/changelog.md#20191213)
- [20191128 - Page texts localization](/docs/changelog.md#20191128)
- [20191127 - Simplified error handling](/docs/changelog.md#20191127)
- [20191126 - Simplified flash messages handling](/docs/changelog.md#20191126)
- [20191124 - Embedded translations and form data session store](/docs/changelog.md#20191124)
- [20191123 - Internationalization](/docs/changelog.md#20191123)

## Installation

[TODO: Create database steps]

```shell
$ git clone https://gitlab.com/mikrowezel/backend/granica
$ make run
```

[TODO: additional steps]

**Test**

```shell
$ make test
```

Use `make grc-test` for colored output.

**Run**

```shell
$ make run
./scripts/run.sh
10:22PM INF New handler name=migration-handler
10:22PM INF New handler name=repo-handler
10:22PM INF New handler name=mailer-handler
10:22PM INF Cookie store key value=pVuOO7ZPNBnqTb5o13JsBMOPcPAe4pxY
10:22PM INF Reading template path=layout/base.tmpl
10:22PM INF Reading template path=user/_ctxbar.tmpl
10:22PM INF Reading template path=user/_flash.tmpl
10:22PM INF Reading template path=user/_form.tmpl
10:22PM INF Reading template path=user/_header.tmpl
10:22PM INF Reading template path=user/_item.tmpl
10:22PM INF Reading template path=user/_list.tmpl
10:22PM INF Reading template path=user/_signin.tmpl
10:22PM INF Reading template path=user/_signup.tmpl
10:22PM INF Reading template path=user/edit.tmpl
10:22PM INF Reading template path=user/index.tmpl
10:22PM INF Reading template path=user/initdel.tmpl
10:22PM INF Reading template path=user/new.tmpl
10:22PM INF Reading template path=user/show.tmpl
10:22PM INF Reading template path=user/signin.tmpl
10:22PM INF Reading template path=user/signup.tmpl
10:22PM INF Parsed template set path=user/edit.tmpl
10:22PM INF Parsed template set path=user/signin.tmpl
10:22PM INF Parsed template set path=user/index.tmpl
10:22PM INF Parsed template set path=user/initdel.tmpl
10:22PM INF Parsed template set path=user/new.tmpl
10:22PM INF Parsed template set path=user/show.tmpl
10:22PM INF Parsed template set path=user/signup.tmpl
10:22PM INF Dialing to Postgres host="host=localhost port=5432 user=granica password=granica dbname=granica sslmode=disable"
10:22PM INF Mailer initializated name=mailer-handler
10:22PM INF Postgres connection established
10:22PM INF Repo initializated name=repo-handler
2019/12/19 22:22:00 Migration 'enable_postgis' already applied.
2019/12/19 22:22:00 Migration 'create_users_table' already applied.
2019/12/19 22:22:00 Migration 'create_accounts_table' already applied.
10:22PM INF Migrator initializated name=migration-handler
10:22PM INF JSON REST Server initializing port=:8081
10:22PM INF Web server initializing port=:8080
```

## Deployment

[To be completed]

## Packages

**Worker**

[Auth](pkg/auth/readme.md)

## Helpers

[Supervisord and Gulp](docs/draft/helpers.md)


docs/changelog.md
# Changelog

## <a name="2019128"></a>20191219 - Mailer implemented

Amazon SES mailer.

Confirmation email can be disabled by configuration but still email can be generated for debugging purposes.

## <a name="2019128"></a>20191217 - User signup

Confirmation email and verification link processing still pending.

## <a name="20191217"></a>20191217 - User signin

[User signin](screenshots.md#user-signin)

## <a name="20191216"></a>20191216 - Model validation and flash messages

Model validations moved to service layer.

[Model validation](screenshots.md#model-validation)

[Flash messages](screenshots.md#flash-messages)

## <a name="20191215"></a>20191215 - Form validations

Final intention is to move validations to the service layer or to another one that located at the "same depth".
In that way they could be reused on JSON REST and gRPC endpoints.

## <a name="20191213"></a>20191213 - RESTful actions completed

RESTful actions completed.

## <a name="20191128"></a>20191128 - Page text localization

Service web pages can now present a localized version of their texts depending on the regional configuration of the browser that accesses the service.

The service use the same translation files under '/assets/web/embed/i18n' to create language bundles.

## <a name="20191127"></a>20191127 - Simplified error handling

Common and repetitive operations on handler errors are now generalized in a single one.

## <a name="20191126"></a>20191126 - Simplified flash messages handling

Wrapped response now takes care to append new messages from current action to pending ones stored by other actions before a redirect.

## <a name="20191124"></a>20191124 - Embedded translations and form data session store

Localization files now are embedded in executable as any other resource.

Endpoint helpers to store, retrieve and clear form data accross requests.

Mainly used to avoid having to fill forms again in case of submission errors.

## <a name="20191123"></a>20191123 - Internationalization

Not fully implemented but basically this is how it works.

The I18N middleware currently tries to read the `lang` field submitted in POST, PUT, PATCH request; alternatively that same field value but read from the query string is used for all methods.

If still not found the value to be used to decide the language to be use will be associated to the one reported by `Accept-Language` request HTTP header.

In any other case the default language (English) will be used.

Soon, priority to decide the language will be the lang user chooses at sign-up time and/or another set by user using a gui switch.

A general clean-up is still needed, eventually a large part of the implementation will be moved to web module.

Finally, as with other assets, localization files under `assets/web/embed/i18n` will be embedded directly into the executable. Temporarily at this stage of development they are accessed from the filesytem.


pkg/auth/auth.go
package auth

import (
	"context"
	"fmt"
	"net/http"
	"sync"

	"github.com/nicksnyder/go-i18n/v2/i18n"
	"gitlab.com/mikrowezel/backend/config"
	"gitlab.com/mikrowezel/backend/granica/pkg/auth/jsonrest"
	"gitlab.com/mikrowezel/backend/granica/pkg/auth/service"
	"gitlab.com/mikrowezel/backend/granica/pkg/auth/web"
	logger "gitlab.com/mikrowezel/backend/log"
	svc "gitlab.com/mikrowezel/backend/service"
)

type (
	Auth struct {
		*svc.BaseWorker
		service        *service.Service
		i18nBundle     *i18n.Bundle
		webep          *web.Endpoint
		jsonep         *jsonrest.Endpoint
		WebServer      http.Handler
		JSONRESTServer http.Handler
	}
)

// NewWorker creates a new Auth worker instance.
func NewWorker(ctx context.Context, cfg *config.Config, log *logger.Logger, name string) (*Auth, error) {
	service := service.MakeService(ctx, cfg, log)

	wep, err := web.MakeEndpoint(ctx, cfg, log, service)
	if err != nil {
		return nil, err
	}

	w := &Auth{
		BaseWorker: svc.NewWorker(ctx, cfg, log, "granica-auth-worker"),
		service:    service,
		webep:      wep,
		jsonep:     jsonrest.MakeEndpoint(ctx, cfg, log, service),
	}

	w.AddWebServer()
	w.AddJSONRESTServer()

	return w, nil
}

func (a *Auth) Init() bool {
	rh, err := a.repoHandler()
	if err != nil {
		return false
	}
	a.service.SetRepo(rh)

	mlh, err := a.mailerHandler()
	if err != nil {
		return false
	}
	a.service.SetMailer(mlh)
	return true
}

func (a *Auth) Start() error {
	var wg sync.WaitGroup

	wg.Add(1)
	go func() {
		a.StartWeb()
		wg.Done()
	}()

	wg.Add(1)
	go func() {
		a.StartJSONREST()
		wg.Done()
	}()

	wg.Wait()
	return nil
}

func (a *Auth) StartWeb() error {
	p := a.Cfg().ValOrDef("web.server.port", "8080")
	p = fmt.Sprintf(":%s", p)

	a.Log().Info("Web server initializing", "port", p)

	err := http.ListenAndServe(p, a.WebServer)
	a.Log().Error(err)
	return err
}

func (a *Auth) StartJSONREST() error {
	p := a.Cfg().ValOrDef("jsonrest.server.port", "8081")
	p = fmt.Sprintf(":%s", p)

	a.Log().Info("JSON REST Server initializing", "port", p)

	err := http.ListenAndServe(p, a.JSONRESTServer)
	a.Log().Error(err)
	return err
}
